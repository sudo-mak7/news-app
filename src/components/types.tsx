import { CommentsInterface } from '@common/types/commentsInterface'

export interface AnswerSectionComponentInterface {
  isCollapsed: boolean,
  answers: CommentsInterface[],
  isLoading: boolean
}