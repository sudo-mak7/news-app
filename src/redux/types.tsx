import * as React from 'react'
import { CommentsInterface } from '@common/types/commentsInterface'
import { NewsInterface } from '@common/types/newsInterface'

export interface fetchAnswersInterface {
  id: number,
  setAnswersState: React.Dispatch<React.SetStateAction<CommentsInterface[]>>,
  setIsLoading: React.Dispatch<React.SetStateAction<boolean>>,
  setError: React.Dispatch<React.SetStateAction<any>>
}

export interface CommentsStateInterface {
  comments: CommentsInterface[],
  loading: boolean,
  error: any
}

export interface CurrentNewsStateInterface {
  currentNews: NewsInterface,
  loading: boolean,
  error: any
}

export interface NewsByIdStateInterface {
  id: number
}

export interface NewsIdsStateInterface {
  newsIds: [number][],
  loading: boolean,
  error: any
}

export interface NewsStateInterface {
  news: NewsInterface[],
  loading: boolean,
  error: any
}

export interface PaginationStateInterface {
  currentPage: number,
  currentPageNews: NewsInterface[],
  pagesLeft: number
}