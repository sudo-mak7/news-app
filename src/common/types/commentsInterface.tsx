interface Comments {
  by: string,
  id: number,
  parent: number,
  text: string,
  time: number,
  type: string
  kids: number[],
  dead: boolean,
  deleted: boolean
}

export type CommentsInterface = Partial<Comments>