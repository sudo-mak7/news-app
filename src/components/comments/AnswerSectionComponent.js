import React from 'react'
import { Comment, Loader } from 'semantic-ui-react'
import CommentComponent from './CommentComponent'

const AnswerSectionComponent = ({ isCollapsed, answers, answersLoading }) => {
  const commentsRender = answers.map(a =>
    <CommentComponent
      key={ a.id }
      id={ a.id }
      { ...a }
    />
  )

  return (
    <Comment.Group collapsed={ isCollapsed }>
      { answersLoading
          ? <Loader active inline='centered'/>
          : commentsRender
      }
    </Comment.Group>
  )
}

export default AnswerSectionComponent