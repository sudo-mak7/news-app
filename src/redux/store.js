import { legacy_createStore as createStore, compose, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'
import rootReducer from './reducers/rootReducer'
import { newsWatcher } from './sagas/rootSaga'

const sagaMiddleware = createSagaMiddleware()

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
    : compose

const configureStore = (preloadedState) => createStore(
  rootReducer,
  preloadedState,
  composeEnhancers(applyMiddleware(sagaMiddleware))
)

const store = configureStore({})

sagaMiddleware.run(newsWatcher)

export default store
