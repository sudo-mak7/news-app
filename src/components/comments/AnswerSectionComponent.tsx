import * as React from 'react'
import { Comment, Header, Loader } from 'semantic-ui-react'
import { useAppSelector } from '@redux/reduxHooks'
import { getCommentsErrorSelector } from '@redux/selectors'
import CommentComponent from '@components/comments/CommentComponent'
import { CommentsInterface } from '@common-types/commentsInterface'

interface AnswerSectionComponentInterface {
  isCollapsed: boolean,
  answers: CommentsInterface[],
  isLoading: boolean
}

const AnswerSectionComponent = (
  { isCollapsed, answers, isLoading }: AnswerSectionComponentInterface
): JSX.Element => {
  const commentsLoadingError = useAppSelector(getCommentsErrorSelector)

  const errorMessage = 'Error loading answers :('

  const commentsRender = answers.map((a: CommentsInterface) =>
    <CommentComponent
      key={ a.id }
      id={ a.id }
      { ...a }
    />
  )

  return (
    isCollapsed
      ? null
      : <Comment.Group>
          { isLoading
              ? <Loader active inline='centered'/>
              : commentsLoadingError
                  ? <Header as='h4' textAlign='center'>{ errorMessage }</Header>
                  : commentsRender
          }
        </Comment.Group>
  )
}

export default AnswerSectionComponent