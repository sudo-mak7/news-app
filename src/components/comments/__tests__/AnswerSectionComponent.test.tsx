import * as React from 'react'
import { store } from '@redux/store'
import { Provider } from 'react-redux'
import { render, screen } from '@testing-library/react'
import { AnswerSectionComponentInterface } from '@components/types'
import AnswerSectionComponent from '@components/comments/AnswerSectionComponent'

const answers = [{
    'by': 'author',
    'id': 2921983,
    'kids': [ 2922097, 2922429, 2924562, 2922709, 2922573, 2922140, 2922141 ],
    'parent': 2921506,
    'text': 'Some text',
    'time': 1314211127,
    'type': 'comment'
}]

const answerSectionComponent = (
  { answers, isCollapsed, isLoading }: AnswerSectionComponentInterface
): JSX.Element =>
  <Provider store={ store }>
    <AnswerSectionComponent
      answers={ answers }
      isCollapsed={ isCollapsed }
      isLoading={ isLoading }
    />
  </Provider>

describe('AnswerSectionComponent test', () => {
  test('AnswerSectionComponent should to match snapshot', () => {
    const props = {
      answers,
      isCollapsed: false,
      isLoading: false
    }

    render(answerSectionComponent(props))
    expect(answerSectionComponent(props)).toMatchSnapshot()
  })

  test('renders AnswerSectionComponent when data loaded', () => {
    const props = {
      answers,
      isCollapsed: false,
      isLoading: false
    }

    render(answerSectionComponent(props))

    const answersInAnswerSectionComponent = screen.getByTestId('answers')
    expect(answersInAnswerSectionComponent).toBeInTheDocument()

    const showAnswersButtonInAnswerSectionComponent = screen.getByText('Show answers')
    expect(showAnswersButtonInAnswerSectionComponent).toBeInTheDocument()
  })

  test('renders AnswerSectionComponent when data is loading', () => {
    const props = {
      answers,
      isCollapsed: false,
      isLoading: true
    }

    render(answerSectionComponent(props))

    const loaderInAnswerSectionComponent = screen.getByTestId('loader')
    expect(loaderInAnswerSectionComponent).toBeInTheDocument()
  })
})