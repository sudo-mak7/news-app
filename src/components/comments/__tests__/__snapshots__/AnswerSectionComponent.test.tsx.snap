// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswerSectionComponent test AnswerSectionComponent should to match snapshot 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <AnswerSectionComponent
    answers={
      Array [
        Object {
          "by": "author",
          "id": 2921983,
          "kids": Array [
            2922097,
            2922429,
            2924562,
            2922709,
            2922573,
            2922140,
            2922141,
          ],
          "parent": 2921506,
          "text": "Some text",
          "time": 1314211127,
          "type": "comment",
        },
      ]
    }
    isCollapsed={false}
    isLoading={false}
  />
</Provider>
`;
